articles/watch-your-state.md

```diff
@@ -0,0 +1,49 @@
+
+    This, however, is very inefficient as the bank's server needs to perform several memory lookups for each customer request to fetch the corresponding token. Unfortunately, most web application firewalls (WAFs) today work exactly this way.
+
+    Before discussing smarter ways of handling access tokens, we look at two more examples.
+
+
+2. Packet forwarding in today's Internet: When Alice sends a data packet to Bob, she adds Bob's IP address to the packet header and hands the packet over to her Internet provider. From then on, every router on the path from Alice to Bob extracts Bob's address from the packet to do a database lookup to retrieve information about the next hop, that is, where to forward the packet to. Each router bases its forwarding decision on local *state* that was previously learned from neighboring routers.
+
+    This procedure is not only insecure (as the neighbors' information is not properly authenticated), but also very energy-consuming and inefficient (as millions of packets per second need to be looked up in specialized parallelizable hardware called TCAMs). Good news: A more secure and -- at the same time -- more efficient solution exists.
+
+
+3. Finally, TCP's vulnerability against denial-of-service attacks: In the initial step of the three-way handshake between client and server, the server needs to store information about the client in order to remember the client when it completes the handshake. The storage of even small amounts of *state* can lead to attacks, when huge numbers of clients pretend to connect to the server, but execute only the first step of the three-way handshake. The server cannot know upfront if a client is malicious and thus keeps waiting for each client to come back to complete the handshake. Soon, when many clients connect simultaneously, the server's resources are exhausted and new connections cannot be accepted. Users then perceive the server as unavailable, aka under a denial-of-service attack.
+
+Higher latency, reduced throughput, increased energy consumption, DoS attacks – what is the solution to prevent these design flaws in authentication, packet forwarding, and TCP? All examples have in common a non-optimal handling of state that is frequently consulted as part of mission-critical business processes.
+
+#### Remediation
+
+A crucial insight for remediation is that state should be stored by the service *requester*, not by the service *provider*. Examples include waiting stamps at the post office counter: they are stored by the waiting customer, not by the post department. Except for a single global counter, the post department keeps no state for each customer waiting in line.
```

Comment:
So wie das hier steht, verstehe ich es als allgemeingültige Anweisung welche für alle Gebiete gilt. Ist das die Intention? Weil im Kontext des Textes passt es ja prima und ist eine gute Überleitung. Aber für allgemeingültig bin ich als Leser jetzt grad etwas überfordert (was heisst das im Kontext von Mobile Applikationen, single page apps, embedded software...?).

Reply:
Guter Punk! Es ist in der Tat als allgemeingültige Anweisung zu verstehen. Die effiziente Kryptographie von heute macht dies möglich. Die Implikationen für andere Kontexte würde den Rahmen des Artikels leider sprengen, aber Beispiele könnte man noch viele liefern.